### ISSUE TITLE
Multi-record serializer / decorator?

### ISSUE THREAD

**Issue by gavin-norman-sociomantic** [sociomantic-tsunami/ocean#551]
I've been looking into https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/util/container/map/utils/MapSerializer.d a bit, recently (while working on https://github.com/sociomantic-tsunami/dhtproto/pull/162).

One quirk of `MapSerializer` is that (as far as I understand) it doesn't serialize each record using the normal struct serializer to load structs. (It implements its own struct version converter, for example: https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/util/container/map/utils/MapSerializer.d#L1143.) The reason for this, apparently, is to speed up loading by only saving the struct version once (at the start of the file), rather than storing it along with every record.

An alternative idea, which would allow us to use a standard serializer for this task:
* A decorator that is tied to a single, specific struct version.
* Upon serializing the first struct, it prepends the version number.
* All subsequent structs are not serialized with a version number.

(Note that my knowledge of the `MapSerializer` and the contiguous struct serializer is not super deep. @mathias-baumann-sociomantic / @mihails-strasuns-sociomantic please correct me if any of this is wrong or nonsensical.)

**Comment by mihails-strasuns-sociomantic** [comment]
It would be trivial to do with https://github.com/sociomantic/ocean/issues/931 :)

**Comment by mathias-baumann-sociomantic** [comment]
Now that you mention it.. if I remember correctly the main motivation for the map serializer was that it auto-converts different struct versions somehow and that was a feature we didn't otherwise have before.

The "one struct version only" thing was just an added bonus due to my sometimes unreasonable urge to optimize. I suppose it wasn't much more work at the time.

**Comment by mihails-strasuns-sociomantic** [comment]
With the feature request of this PR - basically it comes to fundamental mistake that current versioned decorator is context-sensitive - its presence and/or meaning is exclusively defined by a DHT channel, and can't be changed without major deployment disruption. So we can only do it by adding new serializer kinds and there is no way for apps to adjust automatically.

**Comment by mathias-baumann-sociomantic** [comment]
there is also another feature you didn't mention: struct hashes. The map serializer not only saves the version but also a hash of the struct layout. If you would change the struct but not the version you would get a hash checksum error.

**Comment by gavin-norman-sociomantic** [comment]
>Now that you mention it.. if I remember correctly the main motivation for the map serializer was that it auto-converts different struct versions somehow and that was a feature we didn't otherwise have before.

Right, that's what I thought.

>The "one struct version only" thing was just an added bonus due to my sometimes unreasonable urge to optimize. I suppose it wasn't much more work at the time.

Ok, good to know there's not some subtle reason why this is vital. Given that, this feature suggested in this issue may not be necessary at all.

>there is also another feature you didn't mention: struct hashes. The map serializer not only saves the version but also a hash of the struct layout. If you would change the struct but not the version you would get a hash checksum error.

Right. This seems more like a job for a unittest (i.e. confirm that a struct with a certain version has a certain hash), rather than a runtime check upon every deserialization.

>With the feature request of this PR - basically it comes to fundamental mistake that current versioned decorator is context-sensitive - its presence and/or meaning is exclusively defined by a DHT channel, and can't be changed without major deployment disruption. So we can only do it by adding new serializer kinds and there is no way for apps to adjust automatically.

Can you rephrase that? I don't really understand.

**Comment by mihails-strasuns-sociomantic** [comment]
> Can you rephrase that? I don't really understand.

We have already working serializer & version decorator that make most of needed work. However, those are extremely specific:

- First byte must be a version. Passing in unversioned data is undefined behavior and has to be handled differently in code.
- After the version there must be a single serialized struct. You can't have multiple structs. You can't have gzip-compressed struct. You can't have dynamic schema. And so on.
- There is no way to specify abstract metadata other than version - adding new kinds of decorators means changing app code in all cases

So your proposal can certainly be implemented but it will be a totally separate utility and again rely on virtue of developer to only call it on data with proper layout. I think this comes up too often to keep going like that.

**Comment by gavin-norman-sociomantic** [comment]
>So your proposal can certainly be implemented but it will be a totally separate utility and again rely on virtue of developer to only call it on data with proper layout. I think this comes up too often to keep going like that.

This is the idea. My suggestion was only based on a feature of the old `MapSerializer`. It's not intended to be used on data stored in the DHT.

**Comment by gavin-norman-sociomantic** [comment]
Anyway, it sounds like it's not really important to implement this (see https://github.com/sociomantic-tsunami/ocean/issues/551#issuecomment-405566463). Was just an idea.

**Comment by gavin-norman-sociomantic** [comment]
Closing. Seems like it's not important.
